File: packages/agentnet-core/src/agentnet_core/sse.py
"""SSE 线路格式助手:SDK 服务端格式化、Registry 代理/Router 客户端解析共用。"""

from __future__ import annotations


def format_event(event: str, data: str) -> str:
    """序列化一个 SSE 事件块。data 必须是单行(JSON 天然满足)。"""
    return f"event: {event}\ndata: {data}\n\n"


class SseParser:
    """逐行解析 SSE 流;一个事件块(空行分隔)结束时返回 (event, data),否则返回 None。"""

    def __init__(self) -> None:
        self._event: str | None = None
        self._data: list[str] = []

    def feed(self, line: str) -> tuple[str, str] | None:
        line = line.rstrip("\r")
        if line == "":
            if self._event is None and not self._data:
                return None
            event = self._event or "message"
            data = "\n".join(self._data)
            self._event, self._data = None, []
            return event, data
        if line.startswith(":"):
            return None  # 注释/心跳行
        if line.startswith("event:"):
            self._event = line[len("event:"):].strip()
        elif line.startswith("data:"):
            self._data.append(line[len("data:"):].lstrip())
        return None

File: web/src/api/sse.ts
// fetch + ReadableStream 手写 SSE 解析(EventSource 无法带 Authorization header)。
// 线路格式:event: {name}\ndata: {单行 JSON}\n\n;以 ":" 开头的行是心跳/注释,丢弃。

import type { SseEvent } from "./types";

export interface SubscribeOptions {
  signal?: AbortSignal;
  onEvent: (ev: SseEvent) => void;
}

// 解析 SSE 文本流,逐事件回调。网络错误会 throw,由调用方处理。
export async function subscribeSSE(
  url: string,
  headers: Record<string, string>,
  opts: SubscribeOptions,
): Promise<void> {
  const resp = await fetch(url, { headers, signal: opts.signal });
  if (!resp.ok || !resp.body) {
    let msg = `订阅事件流失败(HTTP ${resp.status})`;
    try {
      const body = (await resp.json()) as { message?: string };
      if (body.message) msg = body.message;
    } catch {
      // 保持默认 msg
    }
    throw new Error(msg);
  }

  const reader = resp.body.getReader();
  const decoder = new TextDecoder();
  let buffer = "";
  let eventName = "";
  let dataLines: string[] = [];

  const dispatch = () => {
    if (!eventName || dataLines.length === 0) {
      eventName = "";
      dataLines = [];
      return;
    }
    const raw = dataLines.join("\n");
    try {
      const data = JSON.parse(raw);
      opts.onEvent({ event: eventName, data } as SseEvent);
    } catch {
      // 非 JSON data,忽略
    }
    eventName = "";
    dataLines = [];
  };

  const handleLine = (line: string) => {
    if (line === "") {
      dispatch();
      return;
    }
    if (line.startsWith(":")) return; // 心跳/注释
    const idx = line.indexOf(":");
    const field = idx === -1 ? line : line.slice(0, idx);
    let value = idx === -1 ? "" : line.slice(idx + 1);
    if (value.startsWith(" ")) value = value.slice(1);
    if (field === "event") eventName = value;
    else if (field === "data") dataLines.push(value);
  };

  for (;;) {
    const { done, value } = await reader.read();
    if (done) break;
    buffer += decoder.decode(value, { stream: true });
    let nl: number;
    while ((nl = buffer.indexOf("\n")) !== -1) {
      let line = buffer.slice(0, nl);
      buffer = buffer.slice(nl + 1);
      if (line.endsWith("\r")) line = line.slice(0, -1);
      handleLine(line);
    }
  }
  // 流末尾冲刷残留
  if (buffer.length > 0) handleLine(buffer);
  dispatch();
}
